parse/src/lockfiles/mod.rs

```diff
@@ -0,0 +1,185 @@
+use std::fmt::Display;
+use std::path::Path;
+use std::str::FromStr;
+
+pub use csharp::CSProj;
+pub use java::{GradleLock, Pom};
+pub use javascript::{PackageLock, YarnLock};
+use phylum_types::types::package::{PackageDescriptor, PackageType};
+pub use python::{PipFile, Poetry, PyRequirements};
+pub use ruby::GemLock;
+use serde::de::IntoDeserializer;
+use serde::{Deserialize, Serialize};
+use strum::{EnumIter, IntoEnumIterator};
+
+mod csharp;
+mod java;
+mod javascript;
+mod parsers;
+mod python;
+mod ruby;
+
+/// A file format that can be parsed.
+#[derive(
+    Clone, Copy, Debug, Deserialize, EnumIter, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize,
+)]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+pub enum LockFileFormat {
+    Yarn,
+    Npm,
+    Gem,
+    Pip,
+    Pipenv,
+    Poetry,
+    #[serde(rename = "mvn")]
+    #[serde(alias = "maven")]
+    Maven,
+    Gradle,
+    // This is historically called "nuget" but it's actually for MSBuild project files.
+    // Nuget has its own file formats that are not currently supported.
+    #[serde(rename = "nuget")]
+    #[serde(alias = "msbuild")]
+    Msbuild,
+}
+
+impl FromStr for LockFileFormat {
+    type Err = serde::de::value::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        LockFileFormat::deserialize(s.into_deserializer())
+    }
+}
+
+impl Display for LockFileFormat {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.serialize(f)
+    }
+}
+
+impl LockFileFormat {
+    /// Get the corresponding parser for the specified format.
+    pub fn parser(&self) -> &'static dyn Parse {
+        match self {
+            LockFileFormat::Yarn => &YarnLock,
+            LockFileFormat::Npm => &PackageLock,
+            LockFileFormat::Gem => &GemLock,
+            LockFileFormat::Pip => &PyRequirements,
+            LockFileFormat::Pipenv => &PipFile,
+            LockFileFormat::Poetry => &Poetry,
+            LockFileFormat::Maven => &Pom,
+            LockFileFormat::Gradle => &GradleLock,
+            LockFileFormat::Msbuild => &CSProj,
+        }
+    }
+}
+
+pub type ParseResult = anyhow::Result<Vec<PackageDescriptor>>;
+
+pub trait Parse {
+    /// Parse from a string
+    fn parse(&self, data: &str) -> ParseResult;
+
+    /// Indicate the type of file parsed by this parser
+    fn format(&self) -> LockFileFormat;
```

Comment:
Why did you add this function? As far as I can tell, it's only ever used in tests?

Reply:
This function is required for knowing what format was used when the format is autodetected based on the file name unless you want to try casting the `&dyn Parse` back to every known implementation.

Reply:
So the goal is mostly for non-CLI usage? Do you have a usecase for this already lined up or did you just add it to be safe?

Reply:
This is related to roadmap#140. However, it could probably be useful for API as well. There's nothing very important using this `.format()` at the moment, but it can be useful for things like logging `Parsed file requirements-all.txt as pip` consistently regardless of whether the parser was selected based on the name of the file or by trying formats until finding a parser that works.

Maybe it'd be better to just replace `get_path_parser` with `get_path_format` since you can always go from a format to a parser (and that's actually how `get_path_parser` is implemented).

Reply:
Only implementing LockfileFormat->Parser and then having `get_path_format` seems better.